Core/Src/monitor.c

```diff
@@ -14,71 +14,233 @@
-#define MAX_ADC_VAL_12b 4096
-// DEBUG: threshold may need adjusting
-#define PEDAL_DIFF_THRESH 30
-#define PEDAL_FAULT_TIME  500 /* ms */
+#define TSMS_DEBOUNCE_PERIOD 500 /* ms */
 
 static bool tsms = false;
-static bool brake_state = false;
+osMutexId_t tsms_mutex;
 
-osThreadId lv_monitor_handle;
-const osThreadAttr_t lv_monitor_attributes = {
-	.name = "LVMonitor",
-	.stack_size = 32 * 8,
-	.priority = (osPriority_t)osPriorityHigh4,
-};
-
-void vLVMonitor(void *pv_params)
+/**
+ * @brief Read the open cell voltage of the LV batteries and send a CAN message with the result.
+ * 
+ * @param arg Pointer to mpu_t.
+ */
+void read_lv_sense(void *arg)
 {
-	mpu_t *mpu = (mpu_t *)pv_params;
+	mpu_t *mpu = (mpu_t *)arg;
 	fault_data_t fault_data = { .id = LV_MONITOR_FAULT,
 				    .severity = DEFCON5 };
 	can_msg_t msg = { .id = CANID_LV_MONITOR, .len = 4, .data = { 0 } };
 
 	uint32_t v_int;
 
+	read_lv_voltage(mpu, &v_int);
+
+	/* Convert from raw ADC reading to voltage level */
+
+	// scale up then truncate
+	// convert to out of 24 volts
+	// since 12 bits / 4096
+	// Magic number bc idk the resistors on the voltage divider
+	float v_dec = v_int * 8.967;
+
+	// get final voltage
+	v_int = (uint32_t)(v_dec * 10.0);
+
+	memcpy(msg.data, &v_int, msg.len);
+	if (queue_can_msg(msg)) {
+		fault_data.diag =
+			"Failed to send steering LV monitor CAN message";
+		queue_fault(&fault_data);
+	}
+}
+
+/**
+ * @brief Read data from the fuse monitor GPIO expander on the PDU.
+ * 
+ * @param pdu Pointer to pdu_t.
+ */
+void read_fuse_data(void *arg)
+{
+	pdu_t *pdu = (pdu_t *)arg;
+	fault_data_t fault_data = { .id = FUSE_MONITOR_FAULT,
+				    .severity = DEFCON5 };
+	can_msg_t fuse_msg = { .id = CANID_FUSE, .len = 2, .data = { 0 } };
+	uint16_t fuse_buf;
+	bool fuses[MAX_FUSES] = { 0 };
+
+	struct __attribute__((__packed__)) {
+		uint8_t fuse_1;
+		uint8_t fuse_2;
+	} fuse_data;
+
+	fuse_buf = 0;
+
+	if (read_fuses(pdu, fuses)) {
+		fault_data.diag = "Failed to read fuses";
+		queue_fault(&fault_data);
+	}
+
+	for (fuse_t fuse = 0; fuse < MAX_FUSES; fuse++) {
+		fuse_buf |=
+			fuses[fuse]
+			<< fuse; /* Sets the bit at position `fuse` to the state of the fuse */
+	}
+
+	fuse_data.fuse_1 = fuse_buf & 0xFF;
+	fuse_data.fuse_2 = (fuse_buf >> 8) & 0xFF;
+
+	// reverse the bit order
+	fuse_data.fuse_1 = reverse_bits(fuse_data.fuse_1);
+	fuse_data.fuse_2 = reverse_bits(fuse_data.fuse_2);
+
+	memcpy(fuse_msg.data, &fuse_data, fuse_msg.len);
+	if (queue_can_msg(fuse_msg)) {
+		fault_data.diag = "Failed to send CAN message";
+		queue_fault(&fault_data);
+	}
+}
+
+osThreadId_t non_functional_data_thead;
+const osThreadAttr_t non_functional_data_attributes;
+
+void vNonFunctionalDataCollection(void *pv_params)
+{
+	non_func_data_args_t *args = (non_func_data_args_t *)pv_params;
+	mpu_t *mpu = args->mpu;
+	pdu_t *pdu = args->pdu;
+	free(args);
+
 	for (;;) {
-		read_lv_voltage(mpu, &v_int);
+		read_lv_sense(mpu);
+		read_fuse_data(pdu);
 
```

Comment:
just for my own curiosity: say we wanted to change the frequency of fuse reading down the line, how would we do so?

Reply:
I would use the periodic timer style that I used when sending raw pedal data. Also if we wanted to increase the sample rate we would have to decrease the delay on the task.